feat: Update Select component to match new fundamental styles

salarenko opened the pull request:
#### Please provide a link to the associated issue.
Closes #2006, #2286, #2053, #2431, #2464

#### Please provide a brief summary of this pull request.
This PR:
- adapts new styles and structure of Select component enforced by Fiori3 specification
- improves keyboard navigation
- changes interactive behavior of select control
- introduces `isKey` utility to unify key recognition through different browser platforms
- updates documentation
- removes unsupported by Fiori3 features like:
    - `type SelectType = 'noborder' | 'splitborder';`
    - possibility of defining whole new template using `@Input() triggerTemplate`, narrowing it only to select control content.
- removes Select.isOpen `@Input`
- removes Option.viewValue `@Input`
- removes Select.loading `@Input`

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done

Documentation checklist:
- [x] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

InnaAtanasova commented:
@salarenko Is it `Switch` component or `Select` ?

fkolar commented:
The Select implementation is becoming really complex. One thing that I see and I have commented it in slack:

I see some new code, and I am sure @droshev  already told us about the mobile full screen behavior.
```
<fd-select placeholder="Select an option"
                   [(value)]="selectedValue"
                   [mobile]="true"
                   (isOpenChange)="toggleDialog(dialogTemplate, $event)">
            <ng-template #optionsTemplate>
                <fd-select-mobile-list>
                    <fd-option *ngFor="let option of options" [value]="option">
                        {{option}}
                    </fd-option>
                </fd-select-mobile-list>
            </ng-template>
        </fd-select>
```

The construction seem to be little strange to support  full screen mobile behavior. Let’s think also about the develop experience.  Does this mean that whoever is going to use select and know that it will run on mobile needs to do this above fd-select-mobile-list  ?

You have two things:
you have 
- [mobile]=true
- then this new wrapping  component.

If you are using component you probably want to do only `[mobile]` binding but I guess by default it should be adaptive to any device.  We should do this inside the component .

There is too many combination.  I think it should be possible if you explicitly say its for mobile, to handle this internally do developer can still with `fd-option`.

fkolar commented on libs/core/src/lib/select/select.component.ts:
```diff
@@ -263,211 +308,98 @@ export class SelectComponent implements OnChanges, AfterContentInit, OnInit, OnD
-        return option && this._selected && option === this._selected;
+    /** @hidden */
+    private _setSelectedOption(): void {
+        this.selected = this.options.find(option => option.selected)
```
missing semicolon

fkolar commented on libs/core/src/lib/select/select.component.ts:
```diff
@@ -263,211 +308,98 @@ export class SelectComponent implements OnChanges, AfterContentInit, OnInit, OnD
         }
-        this.close();
-        return;
+        this._changeDetectorRef.detectChanges();
```
You want to be careful when using detectChanges vs MarkForCheck. In this situation you are better with markForCheck. Think of situation where on the same level as this component you will for example some error logic to be displayed , so you want mark this  component and parent to be check in the next cycle so both are updated. 


Such excessive use of DetectChanges just make angular busy more than its needed. When you have a page with 50 similar controls and each of them runs detectChanges it will be pretty noisy, instead of marking component once and Angular can nicely update all marked to be checked. And you do use `OnPush`

fkolar commented on libs/core/src/lib/select/option/option.component.ts:
```diff
@@ -51,35 +50,63 @@ export class OptionComponent implements OnInit {
     @Output()
-    readonly selectedChange: EventEmitter<OptionComponent>
-        = new EventEmitter<OptionComponent>();
+    readonly selectedChange: EventEmitter<OptionComponent> = new EventEmitter<OptionComponent>();
```
Who is going to use this ?  This binding does not have much value  when not connected with the parent `select`

salarenko replied:
Yes, sorry this is a leftover from previous implementation. I left it for eventual backwards compatibility, but since this PR brings breaking changes anyway I'm going to remove it.

salarenko commented:
> The Select implementation is becoming really complex. One thing that I see and I have commented it in slack: 
> ...
Hey @fkolar, thanks for your comments,

we had a discussion about our approach to components and their mobile views and we decided rather to show the user how to create mobile view out of existing components than create mobile mode under the hood of the component.

We decided on his approach mostly because we didn't wanted to bring `DialogModule` with every `ControlModule` and rather give the user an option whether he wants to support mobile mode in his application.

Another thing is `DialogComponent` is highly customizable component, and we wanted to escape hiding its API under the layer of another component.

`SelectComponent` and its mobile mode is on of the "edge cases", because some part of the control needs to be displayed outside of the Dialog. After rethinking this example I think I might be able to escape from using this additional wrapper `fd-select-mobile-list`, but I will still need to render partial views defined in `SelectControl` in `DialogModule`.

In most cases I believe our composition is going to look like this [CalendarComponent in mobile mode](https://fundamental-ngx.netlify.app/#/core/calendar#calendar-mobile).

fkolar commented:
> > The Select implementation is becoming really complex. One thing that I see and I have commented it in slack:
> > ...
> 
> Hey @fkolar, thanks for your comments,
> 
> we had a discussion about our approach to components and their mobile views and we decided rather to show the user how to create mobile view out of existing components than create mobile mode under the hood of the component.
> 
> We decided on his approach mostly because we didn't wanted to bring `DialogModule` with every `ControlModule` and rather give the user an option whether he wants to support mobile mode in his application.
> 
> Another thing is `DialogComponent` is highly customizable component, and we wanted to escape hiding its API under the layer of another component.
> 
> `SelectComponent` and its mobile mode is on of the "edge cases", because some part of the control needs to be displayed outside of the Dialog. After rethinking this example I think I might be able to escape from using this additional wrapper `fd-select-mobile-list`, but I will still need to render partial views defined in `SelectControl` in `DialogModule`.
> 
> In most cases I believe our composition is going to look like this [CalendarComponent in mobile mode](https://fundamental-ngx.netlify.app/#/core/calendar#calendar-mobile).

Thanks for your response. We need to figure this out how to get around this under platform then.  This component & ComboBox are for me candidates to write these components from the scratch. 

Btw. do you still need binding `mobile=true` when you can detect this from the content?

salarenko commented:
> > > The Select implementation is becoming really complex. One thing that I see and I have commented it in slack:
> > > ...
> > 
> > 
> > Hey @fkolar, thanks for your comments,
> > we had a discussion about our approach to components and their mobile views and we decided rather to show the user how to create mobile view out of existing components than create mobile mode under the hood of the component.
> > We decided on his approach mostly because we didn't wanted to bring `DialogModule` with every `ControlModule` and rather give the user an option whether he wants to support mobile mode in his application.
> > Another thing is `DialogComponent` is highly customizable component, and we wanted to escape hiding its API under the layer of another component.
> > `SelectComponent` and its mobile mode is on of the "edge cases", because some part of the control needs to be displayed outside of the Dialog. After rethinking this example I think I might be able to escape from using this additional wrapper `fd-select-mobile-list`, but I will still need to render partial views defined in `SelectControl` in `DialogModule`.
> > In most cases I believe our composition is going to look like this [CalendarComponent in mobile mode](https://fundamental-ngx.netlify.app/#/core/calendar#calendar-mobile).
> 
> Thanks for your response. We need to figure this out how to get around this under platform then. This component & ComboBox are for me candidates to write these components from the scratch.
> 
> Btw. do you still need binding `mobile=true` when you can detect this from the content?

Yes, `[mobile]="true"` is needed, because sometimes component needs to react differently to certain actions in mobile mode. Select for example abandons usage of Popover.

InnaAtanasova commented:
I'm also not sure about the `combobox` role. WAI_ARIA specifies that  the `A composite widget containing a single-line textbox and another element such as listbox or grid, that can dynamically pop up to help the user set the value of the textbox.`. In our case we cannot write inside the input field. We use it to trigger the menu and to show the user the selected value.

InnaAtanasova commented:
Also, a PR marked with feat won't be approved without updated Breaking Changes

Vanessa-Cusmich commented on apps/docs/src/app/core/component-docs/select/select-docs.component.html:
```diff
@@ -1,107 +1,130 @@
 </fd-docs-section-title>
-<description>An example of the select component with diffrent types of Border type and ability to add icon.
-</description>
+<description></description>
```
The states of the Select control can be customized by assigning different border colors for each state.

@Michal: The description was removed here but I think it would be good to add one because all the other examples do :)

mikerodonnell89 commented:
Clicked item has different active state here:
![Screen Shot 2020-04-23 at 1 15 02 PM](https://user-images.githubusercontent.com/2471874/80139988-8c94b080-8564-11ea-93f4-f05bc9dcc2be.png)

vs styles:
![Screen Shot 2020-04-23 at 1 14 50 PM](https://user-images.githubusercontent.com/2471874/80140003-91596480-8564-11ea-8964-fe3e08df2215.png)

Vanessa-Cusmich commented on apps/docs/src/app/core/component-docs/select/select-header/select-header.component.html:
```diff
@@ -1,11 +1,7 @@
 <description>
-    The select component was created to emulate the native select element as closely as possible while also providing
+    The Select component is composed of the SelectComponent and the OptionComponent. It was created to emulate the native Select element as closely as possible while also providing
     more extended styling options.
```
I moved the description of the component to the header (instead of the mode title) because you reference it in several descriptions. Can you confirm if it still makes sense?  Thanks :)

Vanessa-Cusmich commented on apps/docs/src/app/core/component-docs/select/select-docs.component.html:
```diff
@@ -1,107 +1,130 @@
-<fd-docs-section-title [id]="'standard'" [componentName]="'select'">
-    Standard Select
+<fd-docs-section-title [id]="'mode'" [componentName]="'select'">
+    Select mMdes
```
Should this be 'Select Modes'?

mikerodonnell89 commented:
First list item is focused after clicking the select - is this intentional?
![Untitled](https://user-images.githubusercontent.com/2471874/80141330-9d462600-8566-11ea-86b5-5e87d48f3ae4.gif)

mikerodonnell89 commented:
Programmatic open/close it broken
![Untitled](https://user-images.githubusercontent.com/2471874/80143896-cc5e9680-856a-11ea-9915-ddf5b119b561.gif)

mikerodonnell89 commented:
Do we need to close the popover when an option is added or removed?  I think it would be good for the user to actually see this happen

mikerodonnell89 commented on libs/core/src/lib/utils/functions/is-key.ts:
```diff
@@ -0,0 +1,27 @@
+        ['Escape', {aliases: ['Escape', 'Esc'], keyCode: 27}],
+        ['Delete', {aliases: ['Delete', 'Del'], keyCode: 46}],
+        ['Enter', {aliases: ['Enter'], keyCode: 13}],
+        ['Tab', {aliases: ['Tab'], keyCode: 9}],
```
Not being able to check for shift+tab could be an issue in the future

JKMarkowski replied:
For `shift/crtl/alt` + any key, there is flag in object keybaord event object
![image](https://user-images.githubusercontent.com/26483208/80200411-e491ec80-8622-11ea-9875-229bd0a2d9fa.png)

JKMarkowski commented on libs/core/src/lib/select/select.component.ts:
```diff
@@ -74,9 +86,17 @@ export class SelectComponent implements OnChanges, AfterContentInit, OnInit, OnD
     @Input()
     isOpen: boolean = false;
```
Is there way, to control open state only programatically, so clicking inside won't trigger this flag?

salarenko replied:
No there is not. I was also wondering what is the correct behavior for such scenario, but in my opinion isOpen should always reflect select open state.

salarenko commented on libs/core/src/lib/select/option/option.component.ts:
```diff
@@ -1,41 +1,50 @@
-    /** @hidden */
-    @HostBinding('class.fd-menu__item')
-    fdMenuItemClass: boolean = true;
+        .fd-list__item[aria-disabled="true"],
+        .fd-list__item.is-disabled,
+        .fd-list__item:disabled {
+            opacity: 0.4;
+            cursor: not-allowed;
+        }
+        .fd-list__item:disabled:focus {
+            outline: none;
+        }
```
Fundamental Styles List component doesn't support list item disabled state styling (at least I couldn't find it in `v0.8.1`). I added the styles temporarily to NGX until we bring this in Fundamental Styles.

salarenko commented on libs/core/src/lib/select/option/option.component.ts:
```diff
@@ -49,38 +58,61 @@ export class OptionComponent implements OnInit {
+    selected: boolean = false;
+
+    /** @hidden */
+    private _subscriptions: Subscription = new Subscription();
```
This is one Subscription instance which i use as an aggregate for other subscription cancellation functions using `add()` and call all of them with `unsubscribe()` when component is destroyed.

fkolar commented:
> I'm also not sure about the `combobox` role. WAI_ARIA specifies that the `A composite widget containing a single-line textbox and another element such as listbox or grid, that can dynamically pop up to help the user set the value of the textbox.`. In our case we cannot write inside the input field. We use it to trigger the menu and to show the user the selected value.

For Aria the best source is to look at material implementation (autocomplete), they have all the required aria attributes, and this should be good source to double check for all of our components.

JKMarkowski commented on libs/core/src/lib/utils/functions/is-key.ts:
```diff
@@ -0,0 +1,27 @@
+            || keyMap.get(key).keyCode === event.keyCode;
+    }
+
+    throw new Error(`Invalid function arguments. Check if "event" is a KeyboardEvent or "key" is defined in keyMap`);
```
Can you throw it only on devMode? I'm not sure if end-user should see this kind of errors

salarenko commented:
@InnaAtanasova 
You are right about aria `combobox`, I changed it to `listbox` and added also `aria-labelledby` and `aria-activedescendant`.

@mikerodonnell89 @InnaAtanasova 
Select popover is closing in Add/Remove options example because of `closeOnOutsideClick="true"` I switched it to `false` for this example.

@mikerodonnell89 
Could you precise how to reproduce this behavior? For me its:
![image](https://user-images.githubusercontent.com/17496353/80201969-3b002a80-8625-11ea-94c8-93a299832ea5.png)

I also tried to play around with "Programmatic open/close is broken" but I cannot break it. Is there some certain way to do it? Did you found some errors in console?

I'm not sure of we should extend `isKey` with combination keys functionality. The purpose of this util is unification of key codes for different browsers.

@JKMarkowski 
`isKey` now throws errors only in devMode

stefanoScalzo commented:
I dont think that Select in loading mode should not be clickable

stefanoScalzo commented:
Should we imliment some type of close when we click outside of the options for mobile view, they currently have no way to go back without selecting an option

stefanoScalzo commented:
For removing options, should we give the user the ability to choose the min , same for adding and max

stefanoScalzo commented:
The programatic control is broken because when you click change value it selects randomly and not one after the other

salarenko commented:
@stefanoScalzo 
> I dont think that Select in loading mode should not be clickable

That's a good thing to discuss. To be honest I just went through the Fiori3 specs and it says nothing about Loading Mode so we probably event shouldn't introduce it in NGX Core.

> Should we imliment some type of close when we click outside of the options for mobile view, they currently have no way to go back without selecting an option

This is only an example, user can modify Dialog in any way he wants to. You can use `Esc` to close the Dialog. Anyway I just added close button to the example.

> For removing options, should we give the user the ability to choose the min , same for adding and max

I don't think there is a need of creating upper and lower limits for number of option in select.

> The programatic control is broken because when you click change value it selects randomly and not one after the other

I don't agree with that. Change Value is just changing value. Why should it change to the next value in the options list?

Vanessa-Cusmich commented on apps/docs/src/app/core/component-docs/select/select-docs.component.html:
```diff
@@ -1,9 +1,8 @@
 <description>
-    The Select control supports a wide variety of modes to display your list items. <br/>
-    
+    Following examples presents different modes of the Select control.
```
The following examples present the different modes of the Select control.

Vanessa-Cusmich commented on apps/docs/src/app/core/component-docs/select/select-docs.component.html:
```diff
@@ -1,107 +1,131 @@
-    Note that this can be done through ViewChild as well.
+    The content of the Select control can be customized using a custom template which needs to be referenced in the
+    <code>[controlTemplate]</code> input.<br/>
+    Select provides the following context values to the template:
```
Should we move this sentence to line 32 before the break? It looks weird to me.
<img width="820" alt="Screen Shot 2020-04-27 at 9 18 16 AM" src="https://user-images.githubusercontent.com/54723167/80376591-0def8a00-8868-11ea-97b4-d9de3ca58540.png">

Vanessa-Cusmich commented on apps/docs/src/app/core/component-docs/select/select-docs.component.html:
```diff
@@ -1,107 +1,130 @@
+        <li>
+            The Options template needs to be declared inside the <code>fd-select</code> tags to properly configure element
+            injectors. Use <code>Select.selectOptionsListTemplate</code> to render options in the desired place.
+        </li>
```
Should we add breaks between the bullet points to improve readability?

<img width="954" alt="Screen Shot 2020-04-27 at 9 29 34 AM" src="https://user-images.githubusercontent.com/54723167/80378390-9707c080-886a-11ea-8f0e-fb47772ea9ac.png">

stefanoScalzo commented:
> @stefanoScalzo
> 
> > I dont think that Select in loading mode should not be clickable
> 
> That's a good thing to discuss. To be honest I just went through the Fiori3 specs and it says nothing about Loading Mode so we probably event shouldn't introduce it in NGX Core.
>  Perfect ill put it in the chat
> > Should we imliment some type of close when we click outside of the options for mobile view, they currently have no way to go back without selecting an option
> 
> This is only an example, user can modify Dialog in any way he wants to. You can use `Esc` to close the Dialog. Anyway I just added close button to the example.
> Perfect
> > For removing options, should we give the user the ability to choose the min , same for adding and max
> 
> I don't think there is a need of creating upper and lower limits for number of option in select.
> What if the user only wants the minimum to go down to 2 and not 1
> > The programatic control is broken because when you click change value it selects randomly and not one after the other
> 
> I don't agree with that. Change Value is just changing value. Why should it change to the next value in the options list?

Well, I think the user should put a function for the next choice. Not that it is done completely random

JKMarkowski commented on libs/core/src/lib/select/select.component.ts:
```diff
@@ -142,98 +170,154 @@ export class SelectComponent implements OnChanges, AfterContentInit, OnInit, OnD
+
+    /** @hidden */
+    @HostListener('window:resize')
+    resizeScrollHandler(): void {
+        this.calculatedMaxHeight = window.innerHeight * 0.45;
+    }
```
Could you add this method also to `ngAfterViewInit`? calcualtedMaxHeight is only, when someone resize window

salarenko commented:
@JKMarkowski @rengare @InnaAtanasova @mikerodonnell89 @stefanoScalzo @fkolar 

Could you guys take a looks at changes brought with [this recent commit](https://github.com/SAP/fundamental-ngx/pull/2150/commits/84555c9ecc4bb68ce5578182c0fbced6eb5239ee)?

Changes introduced:
- Reorganize `isKey` utility to `KeyUtility` class with static methods:
    - `isKey` - cross browser key unification
    - `isKeyType(alpa | num)` - cross browser alphabetical / numerical key recognition
- Add to `KeyUtility.isKey` possibility to pass collection of keys to check e.g. `KeyUtility.isKey(event, ['Space', 'Enter'])`
- Add selecting options using `ArrowUp` | `ArrowDown` when select is closed.
![list-keyboard-select](https://user-images.githubusercontent.com/17496353/80705050-01786500-8ae6-11ea-9307-dcd4634e88f2.gif)
- Add selecting options using alphanumeric keys
    - If searched phrase has been typed "quickly" Select search and select element by phrase
    - If subset of options matching query is > 1, Select should:
        - Interact with first option if no other option in subset is currently active
        - Interact with next option in the subset after currently active option in the subset
    - Interaction options:
        - Selection if the Select is closed
![list-alpha-select](https://user-images.githubusercontent.com/17496353/80705327-8794ab80-8ae6-11ea-87f3-c32aa6a70d09.gif)
        - Focus if the select is opened
![list-alpha-select-opened](https://user-images.githubusercontent.com/17496353/80705509-d9d5cc80-8ae6-11ea-9da9-c8dff50c283e.gif)

I recommend testing alphanumerical selection on [this example](https://deploy-preview-2150--fundamental-ngx.netlify.app/#/core/select#select-max-height)

rengare commented on libs/platform/src/lib/components/select/select.component.html:
```diff
@@ -1,29 +1,28 @@
-        >
+
+    <fd-option *ngFor="let item of list ; index as index"
+               [disabled]="item.disabled ? true : false"
```
` [disabled]="item.disabled" `

salarenko replied:
This is platform

rengare commented on libs/core/src/lib/select/select.component.ts:
```diff
@@ -142,98 +171,160 @@ export class SelectComponent implements OnChanges, AfterContentInit, OnInit, OnD
+    selected: OptionComponent;
+
+    /** @hidden */
+    optionsArray: OptionComponent[];
```
`options: OptionComponent[];`

optionsArray name is redundant because you can have many options and an array is a type

rengare commented on libs/core/src/lib/select/select.component.ts:
```diff
@@ -131,8 +154,14 @@ export class SelectComponent implements OnChanges, AfterContentInit, OnInit, OnD
+
+    /** Binds to control aria-labelledby attribute */
+    @Input()
+    labelledby: string;
```
`labelledBy`

rengare commented on libs/core/src/lib/button-group/button-group.component.scss:
```diff
@@ -1,3 +1,3 @@
 // @import '~fundamental-styles/dist/button-group';
 @import '~fundamental-styles/dist/button';
-@import '~fundamental-styles/dist/icon';
+@import "~fundamental-styles/dist/icon";
```
kinda strange that it changed to double quote 
![image](https://user-images.githubusercontent.com/10849982/80716740-4efdcd80-8af8-11ea-9986-edfbc2ea4c01.png)
(btw button-group styles are commented out but I'll handle that)

rengare replied:
`@import '~fundamental-styles/dist/icon';`

salarenko commented:
@rengare I addressed you comments.
I also:
- Removed `Select.[isOpen]`
- Removed `Option.[viewValue]`
- Added `Select.[typeaheadDebounceInterval]`
- Changed `selectViewValue` from getter to Select property updated only when needed

fkolar commented on libs/platform/src/lib/components/select/select.component.ts:
```diff
@@ -77,13 +78,12 @@ export class SelectPlatformComponent extends fdSelect implements OnInit, AfterCo
-        super(cd, rtl);
-        this.fdDropdownClass = false;
+    constructor(private cd: ChangeDetectorRef, elementRef: ElementRef) {
+        super(elementRef, new SelectProxy(), cd);
```
We need change this but this is just a comment. - we need to get rid of this inheritance.

fkolar commented on libs/core/src/lib/select/select-proxy.service.ts:
```diff
@@ -0,0 +1,9 @@
+import { OptionStatusChange } from './select.component';
+
+@Injectable()
+export class SelectProxy {
```
What is the added value moving all this to extra service ?

salarenko replied:
Previous implementation included cyclic dependency:
- Select to Option with `ContentChildren(OptionComponent)`
- Option to Select with element injector `constructor(SelectComponent)`

My first try was implementing a communication between Select an Option based on:
- Select to Option with `ContentChildren(OptionComponent)`
- Select subscribes to `Option.onOptionSelect()` emitter
but once again I couldn't escape change detection error and the code have got more complicated because of access to Select->Option in late change detection phase.

I decided to create communication proxy between Select and Option to escape cyclic dependency.

fkolar commented on libs/core/src/lib/select/select.module.ts:
```diff
@@ -16,10 +15,9 @@ import { BusyIndicatorModule } from '../busy-indicator/busy-indicator.module';
         ButtonModule,
         IconModule,
-        LoadingSpinnerModule
+        ListModule
```
Isn't this too big overhead to import module with all its directives ?

salarenko replied:
Well it is. The only alternatives I see are:
- creating a `List->Select` dedicated module
- implementing list features in Select and abandon `ListModule` import
but both of them seems to introduce additional maintenance. This is a thing to discuss with the team if we want to and how we want to handle it.
